07/robot.py

```diff
@@ -0,0 +1,118 @@
+from random import randrange
+
+
+class Robot():
+    max_damage = 5
+
+    def __init__(self, lifes):
+        self.lifes = lifes
+
+    def is_alive(self):
+        """Returns True if robot has any lifes left."""
+        return self.lifes > 0
+
+    def defend(self, damage, attacking_robot):
+        """Default defend. Simply takes damage."""
+        self._take_damage(damage)
+
+    def attack(self, target_robot):
+        """Default attack. Simply makes damage."""
+        self._make_damage(target_robot)
+
+    def _make_damage(self, target_robot):
+        """Generates random damage and forces the
+        other robot to defend."""
+        damage = randrange(0, self.max_damage) + 1
+        target_robot.defend(damage, self)
+
+    def _take_damage(self, damage):
+        """Sets new number of lifes when not already 0."""
+        if self.lifes - damage < 0:
+            self.lifes = 0
+        else:
+            self.lifes -= damage
+
+
+class Aggressive(Robot):
+    max_damage = 7
+
+    def attack(self, target_robot):
+        """Makes damage to the other robot twice.
+        Special for aggresive robot."""
+        self._make_damage(target_robot)
+        self._make_damage(target_robot)
+
+    def get_greeting(self):
+        return 'I will kill you!!! Grrrr!!!'
+
+
+class Defensive(Robot):
+    max_damage = 3
+
+    def defend(self, damage):
+        """Special for deffensive robot, takes only half damage."""
+        self._take_damage(damage // 2)
+
+    def get_greeting(self):
+        return 'Peace!'
+
+
+class Shielded(Robot):
+    def __init__(self, lifes, shields):
+        """Value of shields reflects shields coverage of robot.
+            Number from 0 to 1."""
+        self.shields = shields
+        super().__init__(lifes)
+
+    def defend(self, damage, attacking_robot):
+        """Special for shielded robot, reflects attack to attacking robot."""
+        reflected = int(damage * self.shields)
+        absorbed_by_shields = int(damage * self.shields // 2)
+        self._take_damage(damage - reflected - absorbed_by_shields)
+        if reflected > 0:
+            attacking_robot.defend(reflected, self)
+
+    def get_greeting(self):
+        return 'You have no chance!'
+
+
+class Zombie(Robot):
+    _is_alive = True
+
+    def is_alive(self):
+        return self._is_alive
+
+    def get_greeting(self):
+        return 'Ghhhhhh!'
+
+    def defend(self, damage, attacking_robot):
+        """Special for zombie robot to kill him. """
+        self._take_damage(damage)
+        if self.lifes <= 0 and damage == attacking_robot.max_damage:
+            self._is_alive = False
```

Comment:
zde se nastavuje atribut objektu, takže se atribut třídy nemění, viz níže:

```
(venv) dragca@kvadr:~/pyladies/07$ python
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from robot import zombie
Zombie: Ghhhhhh!
Aggressive: I will kill you!!! Grrrr!!!
Aggressive: 9
Zombie: 0
Aggressive: 5
Zombie: 0
Aggressive: 2
Zombie: 0
Aggressive won
>>> from robot import Zombie
>>> z1 = Zombie(2)
>>> z2 = Zombie(2)
>>> z1.is_alive()
True
>>> z2.is_alive()
True
>>> z1._is_alive = False
>>> z2.is_alive()
True
>>> z1.is_alive()
False
>>> 
```

Reply:
Ale predelala jsem to podle tveho doporuceni, aspon jsem se k tomu zase na chvili vratila.